File: back/src/main/java/com/linkedu/backend/services/DocumentService.java
package com.linkedu.backend.services;

import com.linkedu.backend.entities.User;
import com.linkedu.backend.entities.documents.*;
import com.linkedu.backend.entities.enums.DocumentStatus;
import com.linkedu.backend.entities.enums.DocumentType;
import com.linkedu.backend.repositories.DocumentRepository;
import com.linkedu.backend.repositories.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional
public class DocumentService {

    private final DocumentRepository documentRepository;
    private final UserRepository userRepository;

    private static final String UPLOAD_DIR = "uploads/documents/";

    // ================= FILE SAVE =================
    private String saveFile(MultipartFile file) throws IOException {

        Path uploadPath = Paths.get(UPLOAD_DIR);

        if (!Files.exists(uploadPath)) {
            Files.createDirectories(uploadPath);
        }

        String fileName = UUID.randomUUID() + "_" + file.getOriginalFilename();
        Path filePath = uploadPath.resolve(fileName);

        Files.copy(file.getInputStream(), filePath, StandardCopyOption.REPLACE_EXISTING);

        return filePath.toString();
    }

    // ================= CV =================
    public CvDocument uploadCv(Long studentId,
                               MultipartFile file,
                               String summary,
                               String experience,
                               String skills) {

        try {
            User student = userRepository.findById(studentId)
                    .orElseThrow(() -> new RuntimeException("Student not found"));

            CvDocument doc = new CvDocument();
            doc.setStudent(student);
            doc.setDocumentType(DocumentType.CV);
            doc.setFileName(file.getOriginalFilename());
            doc.setFilePath(saveFile(file));
            doc.setSummary(summary);
            doc.setExperience(experience);
            doc.setSkills(skills);
            doc.setUploadedAt(LocalDateTime.now());

            return documentRepository.save(doc);

        } catch (IOException e) {
            throw new RuntimeException("CV upload failed", e);
        }
    }

    // ================= PASSPORT =================
    public PassportDocument uploadPassport(Long studentId,
                                           MultipartFile file,
                                           String issueDate,
                                           String expiryDate,
                                           String issuingCountry) {

        try {
            User student = userRepository.findById(studentId)
                    .orElseThrow(() -> new RuntimeException("Student not found"));

            PassportDocument doc = new PassportDocument();
            doc.setStudent(student);
            doc.setDocumentType(DocumentType.PASSPORT);
            doc.setFileName(file.getOriginalFilename());
            doc.setFilePath(saveFile(file));
            doc.setIssueDate(issueDate);
            doc.setExpiryDate(expiryDate);
            doc.setIssuingCountry(issuingCountry);
            doc.setUploadedAt(LocalDateTime.now());

            return documentRepository.save(doc);

        } catch (IOException e) {
            throw new RuntimeException("Passport upload failed", e);
        }
    }

    // ================= ID CARD =================
    public IdCardDocument uploadIdCard(Long studentId,
                                       MultipartFile file,
                                       String numId,
                                       String birthday) {

        try {
            User student = userRepository.findById(studentId)
                    .orElseThrow(() -> new RuntimeException("Student not found"));

            IdCardDocument doc = new IdCardDocument();
            doc.setStudent(student);
            doc.setDocumentType(DocumentType.ID_CARD);
            doc.setFileName(file.getOriginalFilename());
            doc.setFilePath(saveFile(file));
            doc.setNumId(numId);
            doc.setBirthday(birthday);
            doc.setUploadedAt(LocalDateTime.now());

            return documentRepository.save(doc);

        } catch (IOException e) {
            throw new RuntimeException("ID Card upload failed", e);
        }
    }

    // ================= GET DOCUMENTS =================
    public List<Document> getStudentDocuments(Long studentId) {

        User student = userRepository.findById(studentId)
                .orElseThrow(() -> new RuntimeException("Student not found"));

        return documentRepository.findByStudent(student);
    }

    //================ VERIFY DOCUMENT ==================
    public Document verifyDocument(Long documentId,
                                   Long agentId,
                                   DocumentStatus status) {

        Document document = documentRepository.findById(documentId)
                .orElseThrow(() -> new RuntimeException("Document not found"));

        User student = document.getStudent();

        if (student.getAssignedAgent() == null) {
            throw new RuntimeException("Student has no assigned agent");
        }

        if (!student.getAssignedAgent().getId().equals(agentId)) {
            throw new RuntimeException("You are not authorized to verify this document");
        }

        if (document.getStatus() != DocumentStatus.PENDING) {
            throw new RuntimeException("Document already verified");
        }

        document.setStatus(status);
        document.setVerifiedBy(student.getAssignedAgent());

        return documentRepository.save(document);
    }
}

File: froent/src/app/student/Document.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

// ── Models ─────────────────────────────────────────────────────────────────

export type DocumentStatus = 'PENDING' | 'VERIFIED' | 'REJECTED';
export type DocumentType   = 'CV' | 'PASSPORT' | 'ID_CARD';

export interface BackendDocument {
  id: number;
  type: DocumentType;
  status: DocumentStatus;
  fileUrl?: string;
  fileName?: string;
  uploadedAt?: string;
  // CV fields
  summary?: string;
  experience?: string;
  skills?: string;
  // Passport fields
  issueDate?: string;
  expiryDate?: string;
  issuingCountry?: string;
  // ID Card fields
  numId?: string;
  birthday?: string;
  // Relations
  student?: { id: number; name?: string };
  verifiedByAgent?: { id: number; name?: string };
}

export interface VerifyDocumentRequestDTO {
  status: DocumentStatus;
}

// ── Service ────────────────────────────────────────────────────────────────

@Injectable({ providedIn: 'root' })
export class DocumentService {

  private readonly base = 'http://localhost:8080/api/documents';

  constructor(private http: HttpClient) {}

  private authHeaders(): HttpHeaders {
    const token = localStorage.getItem('token');
    return new HttpHeaders({ Authorization: `Bearer ${token}` });
  }

  // POST /api/documents/cv
  uploadCv(
    studentId: number,
    file: File,
    summary?: string,
    experience?: string,
    skills?: string
  ): Observable<BackendDocument> {
    const form = new FormData();
    form.append('studentId', String(studentId));
    form.append('file', file);
    if (summary)    form.append('summary', summary);
    if (experience) form.append('experience', experience);
    if (skills)     form.append('skills', skills);
    return this.http.post<BackendDocument>(`${this.base}/cv`, form, {
      headers: this.authHeaders()
    });
  }

  // POST /api/documents/passport
  uploadPassport(
    studentId: number,
    file: File,
    issueDate: string,
    expiryDate: string,
    issuingCountry: string
  ): Observable<BackendDocument> {
    const form = new FormData();
    form.append('studentId', String(studentId));
    form.append('file', file);
    form.append('issueDate', issueDate);
    form.append('expiryDate', expiryDate);
    form.append('issuingCountry', issuingCountry);
    return this.http.post<BackendDocument>(`${this.base}/passport`, form, {
      headers: this.authHeaders()
    });
  }

  // POST /api/documents/id-card
  uploadIdCard(
    studentId: number,
    file: File,
    numId: string,
    birthday: string
  ): Observable<BackendDocument> {
    const form = new FormData();
    form.append('studentId', String(studentId));
    form.append('file', file);
    form.append('numId', numId);
    form.append('birthday', birthday);
    return this.http.post<BackendDocument>(`${this.base}/id-card`, form, {
      headers: this.authHeaders()
    });
  }

  // GET /api/documents/student/{studentId}
  getStudentDocuments(studentId: number): Observable<BackendDocument[]> {
    return this.http.get<BackendDocument[]>(
      `${this.base}/student/${studentId}`,
      { headers: this.authHeaders() }
    );
  }

  // PUT /api/documents/{documentId}/verify?agentId=
  verifyDocument(
    documentId: number,
    agentId: number,
    status: DocumentStatus
  ): Observable<BackendDocument> {
    const body: VerifyDocumentRequestDTO = { status };
    return this.http.put<BackendDocument>(
      `${this.base}/${documentId}/verify?agentId=${agentId}`,
      body,
      { headers: this.authHeaders() }
    );
  }
}